### PR TITLE
Remove secret key

### LINKED ISSUE #834: What’s the secret key used for?
secret_key : used for cryptography purpose

What is a “cryptography purpose”?

### PR DESCRIPTION
**Pull request by Profpatsch**
Instead of having to “spoil” the config with a secret key (which hinders
uploading to a public repo), we can generate it on application start and put it
into the user cache directory (which should only be readable by the searx user).

Adds `utils.get_secret_app_key` which generates a key if none exists and returns
the existing one otherwise (declarative style).
Also test every codepath of that function.

- [x] Tests run w/o errors
- [x] proxying works with the new generated key

fixes https://github.com/asciimoo/searx/issues/834

### CODE REVIEW

File: searx/webapp.py [hunk 1]
```diff
@@ -265,7 +270,7 @@ def proxify(url):
                                            url.encode('utf-8'),
                                            hashlib.sha256).hexdigest()
 
-    return '{0}?{1}'.format(settings['result_proxy']['url'],
+    return '{0}?{1}'.format(settings['re sult_proxy']['url'],
```
Comment (dalf) [inline searx/webapp.py]: There is an extra space I guess

**Comment by asciimoo** [comment]
I feel a bit overkill to introduce a new dependency for this feature.
Wouldn't be enough to add the ability to use environment variable for `secret_key` which overwrites the value defined in `settings.yml`? You can store it in a file and source it before starting searx.

**Comment by Profpatsch** [comment]
> which overwrites the value defined in settings.yml

The whole point is to remove the  `secret_key`, since it’s sensitive data in the config which is completely unnecessary.

We could of course skip the dependency and implement this part of xdg-directory manually, but I don’t see what harm a pure python implementation of xdg does.

File: searx/webapp.py (hunk 1 above)
Comment (Profpatsch) [reply searx/webapp.py]: Right, that slipped in.

Comment (Profpatsch) [reply searx/webapp.py]: I rebased the line, it should be correct now.

**Comment by kvch** [comment]
I have multiple problems with this PR:

1. You submitted multiple things in a single PR: fixing dependency versions and adding a new feature.

2. You overcomplicated it. Publishing secret keys could be avoidable by setting it as an environment variable. This solution requires less change in the current behaviour then yours. The result would be the same. Security would not be a problem as you install/run searx as a user. There is no need to change `settings.yml`, adding a new dependency, generating a secret key. adding new error classes and so on.

3, Introduction of a new dependency. Not only lengthens the installation of searx, but makes the code harder to maintain, because if the API changes, it should be followed up.

So I would not merge this PR at all, as there is a simpler solutions to this problem.

**Comment by kvch** [comment]
I have multiple problems with this PR:

0. Travis fails.

1. You submitted multiple things in a single PR: fixing dependency versions and adding a new feature.

2. You overcomplicated it. Publishing secret keys could be avoidable by setting it as an environment variable. This solution requires less change in the current behaviour then yours. The result would be the same. Security would not be a problem as you install/run searx as a user. There is no need to change `settings.yml`, adding a new dependency, generating a secret key. adding new error classes and so on.

3. Introduction of a new dependency. Not only lengthens the installation of searx, but makes the code harder to maintain, because if the API changes, it should be followed up.

So I would not merge this PR at all, as there is a simpler solutions to this problem.

**Comment by Profpatsch** [comment]
> 0. Travis fails.

There were minor pep warnings, which I fixed.



> You submitted multiple things in a single PR

I can create a new PR for the dependencies if it is wanted.

> setting it as an environment variable. This solution requires less change in the current behaviour then yours. The result would be the same.

That is not a correct.
There is no change in the practical behaviour, as the config variable is superfluous in any case, since the user cannot influence much by setting it manually. It could even be changed to not persist the key and just create one on program start, which would simplify the patch considerably actually. The new dependency could also be removed again.

**Comment by kvch** [comment]
> That is not a correct.
> There is no change in the practical behaviour, as the config variable is superfluous in any case, since the user cannot influence much by setting it manually. 

Generating a secret key automatically is a huge change compared to the current behaviour, because right now it is done manually by the user. In the solution described above secret key is still set manually by the user, so it requires less change. Your reason is partly true, the line in the `settings.yml` would become unnecessary, but the user would be able to set it.

Your solution fails in case of load balanced searx instances. All of the running searx server would have different secret key. So an URL of a proxified image generated by an instance behind LB, could only be served by the instance which generated it. Thus, if you sent the URL to your friend who is served by a different running server, could not open the image, as that server had a different secret key. While having the secret key set in an environment variable or in `settings.yml` would enable all of the instances to use a single secret key.

**Comment by Profpatsch** [comment]
The load-balancing aspect of a bigger setup is very valid, yes. The example of sharing links with friends reminds me of Windows working groups.

What do you think of making local keys (different one for each run, so it doesn’t have to touch the file system) the default case, optionally grabbing a provided key from the environment?
I still think it should be removed from the config if possible.

**Comment by asciimoo** [comment]
@Profpatsch I still can't see the benefit of removing the secret key from `settings.yml` and disallowing the option of setting the secret key there.

If you place the key to another file, searx needs read permission for that file too. What extra safety is provided by this solution compared to the optional redefinition of settings.yml's key with an environment variable (which is ~5 lines of straight-forward code and backward compatible)?

**Comment by virtadpt** [comment]
It looks as if this would make life much more difficult for those of us who use Searx for its API functionality only.

**Comment by Profpatsch** [comment]
> It looks as if this would make life much more difficult for those of us who use Searx for its API functionality only.

Care to elaborate?

**Comment by virtadpt** [comment]
I wrote a number of interactive bots that use Searx's API to issue search requests.  Having the secret key in a configuration file means it's static, which means that I can put it into the configuration files of my bots.  Making it dynamic and cached makes that more difficult to handle.

While this would be merely annoying to me, there are other people out there who are doing much the same thing in more constrained environments, and they won't be able to read the cached file on disk to get the secret key.  This breaks things completely for them.

**Comment by r14c** [comment]
i would never want the application to generate its own secret key, that's definitely something that should be handled explicitly by the operator or some automation system. secret management is **not** the responsibility of the application. using environment variables to define secrets is a much cleaner solution imo. 

my searx cluster uses a config file that is generated with confd, but i'd be happy if i could set the secrets directly with environment variables.

**Pull request closed without merging**